File: CharAnalysis_2_0_MATLAB/src/CharThreshGlobal.m
function [CharThresh] = CharThreshGlobal(Charcoal, Pretreatment, ...
        PeakAnalysis, site, Results, plotData, bkgSensIn)
% CharThreshGlobal    Calculate global threshold value for charcoal peak ID.
%   [CharThresh] = CharThreshGlobal(Charcoal, Pretreatment,
%       PeakAnalysis, site, Results, plotData, bkgSensIn)
%
%   Determines a threshold value based on the distribution of Cpeak across
%   the entire record.  The noise component of Cpeak is modelled as either
%   a 0-mean Gaussian (cPeak = 1, residuals) or a 1-mean Gaussian
%   (cPeak = 2, ratios), or by a Gaussian mixture model.
%
%   INPUTS
%     Charcoal     : struct containing peak (Cpeak series)
%     Pretreatment : struct  (zoneDiv)
%     PeakAnalysis : struct  (cPeak, threshMethod, threshValues)
%     site         : site name string (for plot titles)
%     Results      : struct  (allFigures)
%     plotData     : scalar flag  1 = draw Figure 2
%                    (explicit argument in v2.0; was global in v1.1)
%     bkgSensIn    : scalar flag  0 = normal run,  1 = called from
%                    bkgCharSensitivity (reuse Figure 2 axis limits for
%                    posThreshBins instead of recomputing from data)
%                    (explicit argument in v2.0; was global in v1.1)
%
%   OUTPUT
%     CharThresh : struct with fields:
%                    possible  : vector of candidate threshold values
%                    pos       : [N x 4] positive threshold matrix
%                    neg       : [N x 1] (or [N x 4]) negative threshold
%                    noisePDF  : noise probability density function
%                    SNI       : signal-to-noise index (scalar)
%                    GOF       : goodness-of-fit placeholder (-999 vector)
%
%   v2.0 changes vs v1.1
%     - plotData   passed as explicit argument; global removed.
%     - bkgSensIn  passed as explicit argument; global removed.
%       In v1.1 the global declaration was commented out but the variable
%       was still being written by bkgCharSensitivity and read here via
%       side effects.  It is now a clean argument in both directions.
%     - hist() replaced by charHistCounts() in the plot block.
%     - Bug fix (silent): in the data-defined threshold bin-lookup loop
%       (threshMethod > 1) the second operand of the closest-bin comparison
%       was PeakAnalysis.threshValues(i) instead of thresh(i), meaning the
%       tiebreak compared a percentile value against a CHAR value.  Both
%       operands now consistently use thresh(i).
%
%   CharAnalysis v2.0  -  Phase 1 modernisation

%% ── LOCAL VARIABLES ───────────────────────────────────────────────────────
zoneDiv     = Pretreatment.zoneDiv;
figPosition = [0.1178  0.2158  0.8648  0.6943];

%% ── DEFINE CANDIDATE THRESHOLD BINS ─────────────────────────────────────
%
%   On a normal run (bkgSensIn == 0) the bins span the full range of Cpeak.
%   During the background sensitivity loop (bkgSensIn == 1) the bins are
%   fixed to the x-axis limits already shown on Figure 2, so that the
%   contour plot in bkgCharSensitivity uses a consistent x-axis across all
%   smoothing windows.

if bkgSensIn == 0
    posThreshBins = min(Charcoal.peak) : ...
                    range(Charcoal.peak) / 250 : ...
                    max(Charcoal.peak);
else
    figure(2);
    xlimIn        = get(gca, 'xlim');
    posThreshBins = xlimIn(1) : range(xlimIn) / 250 : xlimIn(2);
end

CharThresh.possible = posThreshBins;

%% ── USER-DEFINED THRESHOLD (threshMethod == 1) ───────────────────────────

if PeakAnalysis.threshMethod == 1

    CharThresh.pos = ones(length(Charcoal.peak), ...
                          length(PeakAnalysis.threshValues));

    for i = 1:length(PeakAnalysis.threshValues)
        tv   = PeakAnalysis.threshValues(i);
        in1  = find(CharThresh.possible >= tv, 1,        'first');
        in2  = find(CharThresh.possible <= tv, 1,        'last');

        % Closest bin to the requested value
        if abs(CharThresh.possible(in1) - tv) <= ...
           abs(CharThresh.possible(in2) - tv)
            inFinal = in1;
        else
            inFinal = in2;
        end

        CharThresh.pos(:,i) = CharThresh.possible(inFinal);
    end

    CharThresh.neg      = -99 * ones(length(Charcoal.peak), ...
                                     length(PeakAnalysis.threshValues));
    CharThresh.noisePDF = -99;

end

%% ── DATA-DEFINED THRESHOLD (threshMethod == 2 or 3) ─────────────────────

if PeakAnalysis.threshMethod > 1

    %% Estimate noise distribution parameters (muHat, sigmaHat)

    if PeakAnalysis.threshMethod == 2

        if PeakAnalysis.cPeak == 1
            % Residuals: noise assumed to be a zero-mean Gaussian.
            % Estimate sigma from the negative half of Cpeak (mirrored).
            negVals  = Charcoal.peak(Charcoal.peak <= 0);
            sigmaHat = std([negVals; abs(negVals)]);
            muHat    = 0;
        else
            % Ratios: noise assumed to be a one-mean Gaussian.
            % Shift values so the distribution is centred on 0, mirror,
            % then shift back.
            subVals  = Charcoal.peak(Charcoal.peak <= 1) - 1;
            sigmaHat = std([subVals; abs(subVals)] + 1);
            muHat    = 1;
        end

        CharThresh.noisePDF = normpdf(posThreshBins, muHat, sigmaHat);

    end

    if PeakAnalysis.threshMethod == 3

        % Gaussian mixture model (two-component)
        [mu, sig] = GaussianMixture(Charcoal.peak, 2, 2, false);

        if mu(1) == mu(2)
            beep
            disp('WARNING: poor fit of Gaussian mixture model.')
            disp('         Re-fitting starting with three classes.')
            [mu, sig] = GaussianMixture(Charcoal.peak, 3, 2, false);
        end

        % Noise component = the Gaussian with the smaller mean
        noiseIdx = find(mu == min(mu), 1);
        sigmaHat = sig(noiseIdx);
        muHat    = mu(noiseIdx);

        CharThresh.noisePDF = normpdf(posThreshBins, muHat, sigmaHat);

    end

    %% Map percentile thresholds to the nearest bin in posThreshBins

    thresh = norminv(PeakAnalysis.threshValues, muHat, sigmaHat);

    CharThresh.pos = ones(length(Charcoal.peak), ...
                          length(PeakAnalysis.threshValues));

    for i = 1:length(PeakAnalysis.threshValues)
        in1 = find(CharThresh.possible >= thresh(i), 1,        'first');
        in2 = find(CharThresh.possible <= thresh(i), 1,        'last');

        % v2.0 bug fix: both operands now use thresh(i).
        % v1.1 used PeakAnalysis.threshValues(i) on the right-hand side,
        % which compared a percentile (e.g. 0.95) against a CHAR value.
        if abs(CharThresh.possible(in1) - thresh(i)) <= ...
           abs(CharThresh.possible(in2) - thresh(i))
            inFinal = in1;
        else
            inFinal = in2;
        end

        CharThresh.pos(:,i) = CharThresh.possible(inFinal);
    end

    % Negative threshold: mirror of the lowest positive percentile
    threshNeg      = norminv(1 - PeakAnalysis.threshValues(3), ...
                             muHat, sigmaHat);
    CharThresh.neg = threshNeg * ones(length(Charcoal.peak), 1);

end

%% ── SIGNAL-TO-NOISE INDEX (Kelly et al.) ────────────────────────────────

signal = Charcoal.peak(Charcoal.peak >  CharThresh.pos(1,4));
noise  = Charcoal.peak(Charcoal.peak <= CharThresh.pos(1,4));

if ~isempty(signal) && length(noise) > 2
    CharThresh.SNI = (1 / length(signal)) .* ...
        sum((signal - mean(noise)) ./ std(noise)) .* ...
        ((length(noise) - 2) / length(noise));
else
    CharThresh.SNI = 0;
end

%% ── GOODNESS-OF-FIT PLACEHOLDER ─────────────────────────────────────────
%   GOF is computed sample-by-sample in CharThreshLocal; for the global
%   threshold a single KS test is not applied, so this field is filled with
%   a sentinel value for downstream compatibility.

CharThresh.GOF = -999 * ones(size(Charcoal.peak));

%% ── PLOT FIGURE 2 ────────────────────────────────────────────────────────

if plotData == 1 && Results.allFigures == 1

    figure(2); clf;
    set(gcf, 'color', 'white', ...
             'name',  ['Peak CHAR distribution, threshold values, ' ...
                       'and noise distribution (if selected)'], ...
             'units', 'normalized', ...
             'position', figPosition)

    % Histogram of Cpeak using charHistCounts (replaces hist())
    [n, x] = charHistCounts(Charcoal.peak, CharThresh.possible);
    h = bar(x, n/sum(n), 1);
    set(h, 'faceColor', [.5 .5 .5], 'edgeColor', [.5 .5 .5])
    hold on

    % Noise PDF overlay (only when data-defined threshold is used)
    if PeakAnalysis.threshMethod > 1
        plot(CharThresh.possible, ...
             CharThresh.noisePDF * mean(diff(CharThresh.possible)), ...
             'k', 'lineWidth', 2)
    end

    % All candidate threshold levels (dashed), selected level (solid red)
    for t = 1:4
        plot([CharThresh.pos(1,t)  CharThresh.pos(1,t)], ...
             [0  max(n/sum(n))], '--k')
    end
    plot([CharThresh.pos(1,4)  CharThresh.pos(1,4)], ...
         [0  max(n/sum(n))], 'r')

    set(gca, 'tickdir', 'out', 'box', 'off')
    ylabel('proportion or scaled density')
    xlabel('peak CHAR (# cm^-^2 yr^-^1)')
    title([char(site) ':  ' num2str(zoneDiv(1)) ' to ' ...
           num2str(zoneDiv(end)) ' cal. yr BP'])

    % Annotation: threshold value and SNI
    xAnnot = CharThresh.pos(1,4);
    yAnnot = max(n/sum(n));

    if PeakAnalysis.threshMethod == 1
        legend('peak CHAR distribution', ...
               'possible threshold values', ...
               'selected threshold value')
        text(1.5*xAnnot, 0.75*yAnnot, ...
             ['Threshold = ' num2str(xAnnot)], ...
             'backgroundColor', 'w')
        text(1.75*xAnnot, 0.60*yAnnot, ...
             ['signal-to-noise index = ' num2str(CharThresh.SNI)], ...
             'backgroundColor', 'w')

    elseif PeakAnalysis.threshMethod == 2 && PeakAnalysis.cPeak == 1
        legend('peak CHAR distribution', ...
               'estimated noise PDF, via 0-mean Gaussian', ...
               'possible threshold values', ...
               'selected threshold value')
        text(1.5*xAnnot, 0.75*yAnnot, ...
             ['Threshold = ' num2str(xAnnot) ';  ' ...
              num2str(round(PeakAnalysis.threshValues(4)*100)) ...
              '^t^h percentile'], ...
             'backgroundColor', 'w')
        text(1.75*xAnnot, 0.60*yAnnot, ...
             ['signal-to-noise index = ' num2str(CharThresh.SNI)], ...
             'backgroundColor', 'w')

    elseif PeakAnalysis.threshMethod == 2 && PeakAnalysis.cPeak == 2
        legend('peak CHAR distribution', ...
               'estimated noise PDF, via 1-mean Gaussian', ...
               'possible threshold values', ...
               'selected threshold value')
        text(1.5*xAnnot, 0.75*yAnnot, ...
             ['Threshold = ' num2str(xAnnot) ';  ' ...
              num2str(round(PeakAnalysis.threshValues(4)*100)) ...
              '^t^h percentile'], ...
             'backgroundColor', 'w')
        text(1.75*xAnnot, 0.60*yAnnot, ...
             ['signal-to-noise index = ' num2str(CharThresh.SNI)], ...
             'backgroundColor', 'w')

    elseif PeakAnalysis.threshMethod == 3
        legend('peak CHAR distribution', ...
               'estimated noise PDF, via Gaussian mixture model', ...
               'possible threshold values', ...
               'selected threshold value')
        text(1.5*xAnnot, 0.75*yAnnot, ...
             ['Threshold = ' num2str(xAnnot) ';  ' ...
              num2str(round(PeakAnalysis.threshValues(4)*100)) ...
              '^t^h percentile'], ...
             'backgroundColor', 'w')
        text(1.75*xAnnot, 0.60*yAnnot, ...
             ['signal-to-noise index = ' num2str(CharThresh.SNI)], ...
             'backgroundColor', 'w')
    end

end

end

File: CharAnalysis_2_0_R/R/char_thresh_global.R
#' Calculate a global threshold value for charcoal peak identification
#'
#' Determines a single threshold applied uniformly across the entire record.
#' The noise component of C_peak is modelled as a zero-mean Gaussian
#' (residuals), a one-mean Gaussian (ratios), or via a Gaussian mixture
#' model (GMM).  Mirrors \code{CharThreshGlobal.m} from the MATLAB v2.0
#' codebase.
#'
#' @param charcoal     Named list containing \code{peak} (C_peak series) and
#'   \code{ybpI}.
#' @param pretreatment Named list with \code{zoneDiv}.
#' @param peak_analysis Named list with \code{cPeak}, \code{threshMethod},
#'   \code{threshValues}.
#' @param site         Character string (site name; unused in R, kept for API
#'   symmetry).
#' @param results      Named list (unused in R, kept for API symmetry).
#' @param plot_data    0/1 flag; ignored in R.
#' @param bkg_sens_in  0/1 flag; ignored in R (no sensitivity loop).
#'
#' @return Named list \code{char_thresh} with elements:
#'   \describe{
#'     \item{possible}{Numeric vector of 251 candidate threshold bins.}
#'     \item{pos}{Numeric matrix [N x 4]: positive threshold for each of the
#'       four \code{threshValues}.}
#'     \item{neg}{Numeric matrix [N x 4] (method 1) or [N x 1] (methods 2–3):
#'       negative threshold.}
#'     \item{noise_pdf}{Estimated noise PDF evaluated at \code{possible}
#'       (methods 2–3), or scalar \code{-99} (method 1).}
#'     \item{mu_hat}{Fitted noise-component mean.}
#'     \item{sigma_hat}{Fitted noise-component standard deviation.}
#'     \item{SNI}{Signal-to-noise index (scalar).}
#'     \item{GOF}{Sentinel vector (\code{-999}, length N).}
#'   }
#'
#' @details
#'   ## Gaussian assumption (threshMethod = 2)
#'   For residuals (\code{cPeak = 1}): noise is zero-mean; sigma is estimated
#'   from the negative half of C_peak, mirrored and pooled.
#'   For ratios (\code{cPeak = 2}): noise is one-mean; values are shifted to
#'   zero, mirrored, shifted back, and sigma estimated from the pooled set.
#'
#'   ## GMM (threshMethod = 3)
#'   MATLAB's \code{GaussianMixture(X, 2, 2, false)} is replicated by
#'   \code{gaussian_mixture_em(X)} from \code{utils_gaussian_mixture.R}, using
#'   the same first/last-point initialisation and loose convergence criterion as
#'   the original Bowman CLUSTER EM.  The noise component is identified as the
#'   Gaussian with the smaller mean (matching MATLAB's
#'   \code{noiseIdx = find(mu == min(mu), 1)}).
#'
#'   ## Bin-lookup for threshold values
#'   Percentile thresholds are mapped to the nearest bin in
#'   \code{possible} (251 equally-spaced values spanning C_peak range).
#'   The v2.0 bug fix is preserved: both sides of the \code{abs()} comparison
#'   use the CHAR-unit threshold value \code{thresh[i]}, not the raw percentile.
#'
#' @seealso [char_thresh_local()], [char_smooth()], [CharAnalysis()]

# Requires: gaussian_mixture_em() from utils_gaussian_mixture.R
# (source that file before sourcing this one, or ensure it is on the search path)

char_thresh_global <- function(charcoal, pretreatment, peak_analysis,
                                site = NULL, results = NULL,
                                plot_data = 0L, bkg_sens_in = 0L) {

  # Candidate threshold bins (251 values spanning the C_peak range)
  pos_thresh_bins <- seq(min(charcoal$peak, na.rm = TRUE),
                         max(charcoal$peak, na.rm = TRUE),
                         length.out = 251L)

  char_thresh <- list(possible = pos_thresh_bins)
  N           <- length(charcoal$peak)
  n_tv        <- length(peak_analysis$threshValues)

  mu_hat    <- NA_real_
  sigma_hat <- NA_real_

  # =========================================================================
  # METHOD 1: User-defined threshold
  # =========================================================================
  if (peak_analysis$threshMethod == 1L) {

    char_thresh$pos <- matrix(NA_real_, nrow = N, ncol = n_tv)

    for (i in seq_len(n_tv)) {
      tv    <- peak_analysis$threshValues[i]
      in1   <- which(pos_thresh_bins >= tv)[1L]
      in2   <- rev(which(pos_thresh_bins <= tv))[1L]
      in1   <- if (is.na(in1)) length(pos_thresh_bins) else in1
      in2   <- if (is.na(in2)) 1L                       else in2
      in_final <- if (abs(pos_thresh_bins[in1] - tv) <=
                      abs(pos_thresh_bins[in2] - tv)) in1 else in2
      char_thresh$pos[, i] <- pos_thresh_bins[in_final]
    }

    char_thresh$neg       <- matrix(-99, nrow = N, ncol = n_tv)
    char_thresh$noise_pdf <- -99

  }

  # =========================================================================
  # METHODS 2 AND 3: Data-defined threshold
  # =========================================================================
  if (peak_analysis$threshMethod > 1L) {

    # -- Estimate noise distribution ----------------------------------------
    if (peak_analysis$threshMethod == 2L) {

      if (peak_analysis$cPeak == 1L) {
        # Residuals: zero-mean Gaussian
        neg_vals  <- charcoal$peak[charcoal$peak <= 0 & !is.na(charcoal$peak)]
        sigma_hat <- stats::sd(c(neg_vals, abs(neg_vals)))
        mu_hat    <- 0
      } else {
        # Ratios: one-mean Gaussian
        sub_vals  <- charcoal$peak[charcoal$peak <= 1 &
                                     !is.na(charcoal$peak)] - 1
        sigma_hat <- stats::sd(c(sub_vals, abs(sub_vals)) + 1)
        mu_hat    <- 1
      }

      char_thresh$noise_pdf <- stats::dnorm(pos_thresh_bins, mu_hat, sigma_hat)
    }

    if (peak_analysis$threshMethod == 3L) {

      # GMM: replicate MATLAB's GaussianMixture(X, 2, 2, false) using the
      # direct R port gaussian_mixture_em() from utils_gaussian_mixture.R.
      gmm_data <- charcoal$peak[!is.na(charcoal$peak)]
      fit      <- tryCatch(
        gaussian_mixture_em(gmm_data, k = 2L),
        error = function(e) NULL
      )

      if (is.null(fit)) {
        # EM failed: fall back to simple Gaussian (method 2 behaviour)
        if (peak_analysis$cPeak == 1L) {
          neg_vals  <- gmm_data[gmm_data <= 0]
          sigma_hat <- stats::sd(c(neg_vals, abs(neg_vals)))
          mu_hat    <- 0
        } else {
          sub_vals  <- gmm_data[gmm_data <= 1] - 1
          sigma_hat <- stats::sd(c(sub_vals, abs(sub_vals)) + 1)
          mu_hat    <- 1
        }
      } else {
        # gaussian_mixture_em returns mu and sigma sorted ascending
        mu_both    <- fit$mu
        sigma_both <- fit$sigma

        if (mu_both[1L] == mu_both[2L]) {
          # Poor GMM fit — re-fit with K = 3, take two smallest-mean components
          warning("char_thresh_global: poor GMM fit (mu1 == mu2). ",
                  "Re-fitting with K = 3.")
          fit3 <- tryCatch(
            gaussian_mixture_em(gmm_data, k = 3L),
            error = function(e) NULL
          )
          if (!is.null(fit3)) {
            mu_both    <- fit3$mu[1:2]
            sigma_both <- fit3$sigma[1:2]
          }
        }

        # Noise component = Gaussian with the smaller mean (index 1, sorted)
        mu_hat    <- mu_both[1L]
        sigma_hat <- sigma_both[1L]
      }

      char_thresh$noise_pdf <- stats::dnorm(pos_thresh_bins, mu_hat, sigma_hat)
    }

    # -- Map percentile thresholds to nearest bin ----------------------------
    thresh_vals <- stats::qnorm(peak_analysis$threshValues, mu_hat, sigma_hat)

    char_thresh$pos <- matrix(NA_real_, nrow = N, ncol = n_tv)

    for (i in seq_len(n_tv)) {
      tv    <- thresh_vals[i]
      in1   <- which(pos_thresh_bins >= tv)[1L]
      in2   <- rev(which(pos_thresh_bins <= tv))[1L]
      in1   <- if (is.na(in1)) length(pos_thresh_bins) else in1
      in2   <- if (is.na(in2)) 1L                       else in2
      # v2.0 bug fix: both operands use thresh_vals[i], not threshValues[i]
      in_final <- if (abs(pos_thresh_bins[in1] - tv) <=
                      abs(pos_thresh_bins[in2] - tv)) in1 else in2
      char_thresh$pos[, i] <- pos_thresh_bins[in_final]
    }

    # Negative threshold: mirror of the THIRD percentile value
    thresh_neg       <- stats::qnorm(1 - peak_analysis$threshValues[3L],
                                      mu_hat, sigma_hat)
    char_thresh$neg  <- matrix(thresh_neg, nrow = N, ncol = 1L)
  }

  # =========================================================================
  # SIGNAL-TO-NOISE INDEX (Kelly et al.)
  # =========================================================================
  t_pos  <- char_thresh$pos[1L, n_tv]    # threshold using final threshValue
  signal <- charcoal$peak[!is.na(charcoal$peak) & charcoal$peak >  t_pos]
  noise  <- charcoal$peak[!is.na(charcoal$peak) & charcoal$peak <= t_pos]

  char_thresh$SNI <- if (length(signal) > 0 && length(noise) > 2) {
    (1 / length(signal)) *
      sum((signal - mean(noise)) / stats::sd(noise)) *
      ((length(noise) - 2) / length(noise))
  } else {
    0
  }

  # =========================================================================
  # GOODNESS-OF-FIT PLACEHOLDER
  # For the global threshold a single KS test is not applied; fill with the
  # sentinel value -999 for downstream compatibility.
  # =========================================================================
  char_thresh$GOF <- rep(-999, N)

  char_thresh$mu_hat    <- mu_hat
  char_thresh$sigma_hat <- sigma_hat

  char_thresh
}
